### ISSUE TITLE
--campaign-limit flag on campaign trigger cron is not working after upgrading to 3.3.1

### ISSUE THREAD

**Issue by jcmartinez** [mautic/mautic#9750]
## Bug Description

Yesterday I updated from 3.2.5 to 3.3.1. I must say, I never updated to 3.3.0 - not sure if this was needed.

I have a marketing campaign scheduled with a large number of subscribers that was supposed to kick in today, but no email message shows in the calendar nor any has been sent.

| Q   | A
| --- | ---
| Mautic version | 3.3.1
| PHP version | 7.3 and 7.4 (tested both)
| Browser | Chrome

### Steps to reproduce
1. Create a segment with a arge number of contacts.
2. Create a new email using the blank template. This email was cloned from a previously sent message that worked..
3. Create a very basic marketing campaign targeting the segment, at a relative time, during business hours. This campaign adds 10 points to the contact if it logs into the site. This campaign was cloned from a previous campaign that worked.
 
### Log errors

When I run:

```
sudo -u www-data php web/list/bin/console mautic:campaigns:trigger --campaign-limit=50
[sudo] password for myusername: 
Triggering events for campaign 6
Triggering events for newly added contacts
0 total events(s) to be processed in batches of 100 contacts

0 total events were executed
0 total events were scheduled

Triggering scheduled events
0 total events(s) to be processed in batches of 100 contacts

0 total events were executed
0 total events were scheduled

Triggering events for inactive contacts

0 total events were executed
0 total events were scheduled

Triggering events for campaign 8
Triggering events for newly added contacts
16071 total events(s) to be processed in batches of 100 contacts
 16071/16071 [============================] 100%
```

... I see the following logs being added:

```
[2021-03-02 17:56:53] mautic.ERROR: CAMPAIGN: Campaign Limit exceeded [] []
[2021-03-02 18:53:31] mautic.ERROR: CAMPAIGN: Campaign Limit exceeded [] []
```

**Comment by RCheesley** [comment]
Looking back through the commits that introduced this feature it seems to be related to

> limit the number of contacts to process per campaign, so that no one campaign causes a substantial delay to all other campaigns

Are your cron jobs setting a limit using

>  --campaign-limit=XXX option with bin/console mautic:campaigns:trigger command

That would cause the error mentioned!

**Comment by jcmartinez** [comment]
@RCheesley - Thank you for looking into this so quickly.

Yes, I'm using a limit because I need to stay below a specific hourly threshold.
```
php web/list/bin/console mautic:campaigns:trigger --campaign-limit=50
```
Before v3.3.1, the above cron used to put 50 messages per run in the var/spool/ directory. Then the following cron would send these 50 email messages out in smaller batches:
```
php web/list/bin/console mautic:emails:send --message-limit=10
```
With this, I had a steady flow of email messages going out. I feel like this was working very well in combination with the scheduling settings of the campaign. My goal was to send these email messages only during the subscriber's business hours and never going above a specific number of messages per hour.

Did the commit you mentioned introduced a bug or should I use a different method to deliver the specific number of messages per hour within the configured schedule of the campaign?

**Comment by RCheesley** [comment]
I was referring to the years-old commit that introduced the limit feature - I had not come across it before!  However the issue that you are getting seems to be that you have already hit that limit, hence the message in the logs.

Maybe both campaigns are now running at the same time and you are hitting the limit? Are any emails going out at all?

**Comment by jcmartinez** [comment]
No, no email is going out at all. No email is being placed into the var/spool/ directory.

Does this mean that I should have only one campaign running at a time?

I've been keeping a previous campaign active, even though it already finished sending. Not sure whether this is necessary for users to unsubscribe, visit the links, clicks being tracked, etc. If not, I will stop it.

I upgraded straight from v3.2.5 to v3.3.1, so I never was on v3.3.0. Not sure if that matters.

My dashboard didn't work initially, even though I flushed cache multiple times. The dashboard started working the next day. Could it be that I'm running with a broken upgrade?

**Comment by jcmartinez** [comment]
I just disabled the older campaign that had finished running leaving only the new one. I still get the same error in the logs.

**Comment by jcmartinez** [comment]
This issue is happening on Mautic 3.3.1 on a fresh install.

Steps to reproduce on a fresh installation:

1. Install a fresh copy of Mautic 3.3.1, with PHP 7.4, and MariaDB 10.5.9 on a local linux environment
2. import 4000 contacts from CSV
3. create a segment that selects all emails
4. create an email using the Aurora template
5. created a campaign to be executed at a relative time: 0 days, scheduled only: on Weekdays, Email to send: the one created, Email type: Marketing.
6. Set enabled the campaign, the email, the segment.
7. Set mautic to send emails using the queue
8. Run the following commands:
```
php /app/web/list/bin/console mautic:segments:update
php /app/web/list/bin/console mautic:campaigns:rebuild
php /app/web/list/bin/console mautic:campaigns:update
php /app/web/list/bin/console mautic:campaigns:trigger --campaign-limit=50
```
No email message is generated. 

The logs show:
```
[2021-03-05 21:19:30] mautic.ERROR: CAMPAIGN: Campaign Limit exceeded [] []
```
If you trigger the campaign without the limit, no error is shown, but all 4,000 email messages are generated at the same time and placed into the spool directory. This is not the expected behavior, as it would defeat the purpose of having a schedule configuration for the campaign.

**Comment by RCheesley** [comment]
Thanks for debugging this further, I have renamed the issue so that it is clearer that this is relating to the campaign limit flag. Hopefully it might be possible to find someone to work on this during the next few weeks as our next release is focusing on email!

**Comment by kuzmany** [comment]
@jcmartinez 

please can you check if this reverted code will works for you https://github.com/mautic/mautic/commit/32dbebdf7c2e4a8a866952cb8f5861e813bd376f#diff-cdc167810713fe9774f2fb69e11deed430454d5af018220dd25b3100f172f6f3R250

I mean  replace

`} elseif ($this->campaignLimit <= ($this->campaignLimitUsed + $reduction)) {`

to 

`} elseif ($this->campaignLimit < ($this->campaignLimitUsed + $reduction)) {`

**Comment by kuzmany** [comment]
@jcmartinez fixed https://github.com/mautic/mautic/pull/9756 . Please test

**Comment by RCheesley** [comment]
Resolved in 3.3.2 Release Candidate coming soon! 🎉
